### PR TITLE
Beta history selenium updates and assorted bug fixes

### PR DESCRIPTION
**Pull request by Nerdinacan**
This PR is primarily to make the beta history selenium tests pass.

Most of this is editing the selenium tests themselves though a few minor bugs were exposed by the tests. Changes to the client are mostly the addition of data-description="whatever" attributes so that selenium has something non-invasive to look for, but the testing did expose some bugs as well:

### Notable but minor client changes in order of significance

1. There was a problem with the recursive display of nested dataset collections. An old component placeholder was being delivered instead of a proper ContentItem variant.

2. Simplified the rendering/saving of tags to avoid the old tag dependency injection scheme when saving tags. Now the stateless component emits events up to the editing wrapper and the history/dataset/whatever is saved as a unit.

3. The dataset menu which was recently revamped needed a writable flag to eliminate inappropriate editing menu options for datasets that are part of an immutable collection

4. Corrected state calculation in model/DatasetCollection and model/JobStateSummary. The logic still looks funny to me but it emulates the existing logic from the backbone code. Some cosmetic visual indicators like the spinning wheel have not be implemented, but the logic state is being calculated and delivered properly.

5. Tweaked the code in contentPayload.js to provide a slightly friendlier experience when the user is near the top of the history when new updates are being rendered. If you are in the middle of a long list, and the polling detects updates above the area being viewed, you would not expect the scroller to shift to those updates. However, if you are already at the top, the expectation is that you will see the newest updates arrive. The provider attempts to figure out whether the user is at the top of the list and if the updates are the most recent updates in the history, and if so, adjusts the window to the top.

6. There was a significant and hard-to-find bug in the Rules collection builder modal that stemmed from the fact that the  collection modal component is still using backbone models which are so outdated that they use a deprecated jQuery Deferred object internally instead of a standard Promise. This is a problem because the then() signatures are different.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

**Comment by Nerdinacan** [comment]
@jmchilton 

I am trying to get the following scheme to work but am running into problems related to my unfamiliarity with the SmartComponent framework, But, the most sensible strategy seem to be something like this this:

```yml
history_panel:
  beta_item:
    selectors:
      _: '.beta.history'
      details: '${_} .details'
      title: '${_} .content-title'
```

Then it would be very flexible to be able to add data selectors to the base selector using:

```python
title = self.components.history_panel.beta_item.selector.with_data('hid', 2).title
```

But in that sample "title" is a "SmartTemplate" object which I'm not sure how to change back into a SmartComponent object to get access to all the important .when_preset(), .wait_for_and_click(), etc. methods

**Comment by Nerdinacan** [comment]
Looks like the job state summary tables are not being populated during the tests, thus the view I wrote here:
https://github.com/galaxyproject/galaxy/pull/9713

...is no longer returning the right job_state_summary object in the contents/near api response, which means none of the collection status tests are going to register any intermediary states during job execution.

Perhaps the jobs api was refactored recently?

**Comment by davelopez** [comment]
> Perhaps the jobs api was refactored recently?

Not that I'm aware of, at least the REST API part, but you are right, it looks like something strange is going on. I've just had a look at my local `history_dataset_collection_association` table and all job-related ids are null, but they shouldn't, right?

Screenshot from 2021-08-25 11-55-50

**Comment by Nerdinacan** [comment]
@davelopez 

Got some more details. I split a large text file into a collection using a split tool. If you run it in the legacy UI, a separate set of polling occurs to monitor the specific job. It hits this endpoint to get its job states:

Doing so in the legacy app polls this endpoint to get current jobs:

```
// http://localhost:8081/api/histories/5e4434cba401361d/jobs_summary?ids=c8a1d1385ed48f8d&types=Job

[
  {
    "populated_state": "ok",
    "states": {
      "ok": 1
    },
    "model": "Job",
    "id": "c8a1d1385ed48f8d"
  }
]
```

But later polling in the beta history hits the contents/near endpoint for the same collection, and I note that at no point does the job_state_summary ever get updated.

```
// 20210825091455
// http://localhost:8081/api/histories/5e4434cba401361d/contents/near/1080/60?visible=True&deleted=False

[
  {
    "populated_state_message": null,
    "update_time": "2021-08-25T16:11:37.239337",
    "job_source_type": "Job",
    "type_id": "dataset_collection-b887d74393f85b6d",
    "job_source_id": "c8a1d1385ed48f8d",
    "type": "collection",
    "collection_type": "list",
    "url": "/api/histories/5e4434cba401361d/contents/dataset_collections/b887d74393f85b6d",
    "job_state_summary": {
      "new": 0,
      "waiting": 0,
      "running": 0,
      "ok": 0,
      "failed": 0,
      "deleted": 0,
      "upload": 0,
      "queued": 0,
      "resubmitted": 0,
      "error": 0,
      "paused": 0,
      "deleted_new": 0,
      "all_jobs": 0
    },
    "populated": true,
    "contents_url": "/api/dataset_collections/b887d74393f85b6d/contents/36ddb788a0f14eb3",
    "create_time": "2021-08-25T16:11:27.891635",
    "collection_id": 46,
    "history_content_type": "dataset_collection",
    "hid": 606,
    "populated_state": "ok",
    "deleted": false,
    "history_id": "5e4434cba401361d",
    "id": "b887d74393f85b6d",
    "element_count": 500,
    "name": "Split file on data 3",
    "tags": [
      
    ],
    "visible": true
  },
]
```

Unless something has changed, the job state summary values come from a view in the database:

```sql

select * from collection_job_state_summary_view
```

Which gets its values from the following source tables:

```sql

   blah blah blah

   FROM (((history_dataset_collection_association hdca
     LEFT JOIN implicit_collection_jobs icj ON ((icj.id = hdca.implicit_collection_jobs_id)))
     LEFT JOIN implicit_collection_jobs_job_association icjja ON ((icj.id = icjja.implicit_collection_jobs_id)))
     LEFT JOIN job ON ((icjja.job_id = job.id)))
  GROUP BY hdca.id;

```

So I guess those tables no longer house the job states?

**Comment by Nerdinacan** [comment]
I note that the join is failing because both of the implicit collection tables are completely empty. Maybe they are now vestigial? I also note there is a job_id on the history_dataset_collection_association table, which I vaguely remember a PR for.

I can rewrite this view to use hdca.job_id instead, but I'm surprised they didn't just drop the two implicit jobs tables if they are no longer to be used.

**Comment by Nerdinacan** [comment]
I'm all out of time. I think this covers the tests that actually pertain to the history panel code. There are some tests marked as beta for the multi-history view and the structural views, but those pieces have not been implemented.

**Comment by dannon** [comment]
@Nerdinacan Awesome, I'll take a look!  I saw that there were just a couple left when I checked through last night, thank you.

**Comment by Nerdinacan** [comment]
The most problematic test has been collection builders.

Here's the behavior:

The files upload, fine. Then the HistoryContentProvider renders the new content.

On a real browser, an additional step in contentPayload.js emits an event that adjusts the scroller to get to the top of the list so that the user can see the new stuff that appeared at the top, above the previous window of view. This conditional event is necessary because you wouldn't want this happening if you were just down in the middle of the list someplace and some random stretch of content above you got modified, that's what caused a lot of the jerking before.

On a real browser, the reset event is emitted and the scroller shifts, but with selenium it's a crap shoot. I had to throw in a self.home() to cheat my way out of the test.

File: client/src/components/History/providers/HistoryContentProvider/contentPayload.js [hunk 1]
```diff
@@ -149,18 +152,13 @@ export const contentPayload = (cfg = {}) => {
...
-                // console.groupEnd();
+                const scrollNearLastTop = pos.key !== null ? Math.abs(pos.key - lastResponse.maxContentHid) < fudge : false;
+                const scrollerAtTop = scrollerExactlyAtTop || scrollNearLastTop;
 
                 if (updatesAtTop && scrollerAtTop) {
-                    return new ScrollPos({ cursor: 0, key: response.maxHid })
+                    return ScrollPos.create({ cursor: 0, key: response.maxContentHid });
                 }
                 return null;
             }),
```
Comment (Nerdinacan) [inline client/src/components/History/providers/HistoryContentProvider/contentPayload.js:164]: This is the code responsible for emitting the event that shifts the scroller back to the top after a content update.

**Review by dannon — APPROVED** [review]

**Comment by dannon** [comment]
The test `FAILED lib/galaxy_test/selenium/test_history_panel.py::HistoryPanelTestCase::test_refresh_preserves_state` has been fixed upstream and passes locally on an integration branch, going to go ahead and get this batch in.  Thanks @Nerdinacan

**Comment by dannon** [comment]
Argh, after pulling this in I do see that `lib/galaxy_test/selenium/test_tool_form.py::ToolFormTestCase::test_rerun` is failing locally.  Investigating.

**Pull request merged**